devito/operations/interpolators.py

```diff
@@ -33,6 +34,59 @@ def wrapper(interp, *args, **kwargs):
     return wrapper
 
 
+def extract_subdomain(variables):
+    """
+    Check if any of the variables provided are defined on a SubDomain
+    and extract it if this is the case.
+    """
+    sdms = set()
+    for v in variables:
+        try:
+            if v.grid.is_SubDomain:
+                sdms.add(v.grid)
+        except AttributeError:
+            # Variable not on a grid (Indexed for example)
+            pass
+
+    if len(sdms) > 1:
+        raise NotImplementedError("Sparse operation on multiple Functions defined on"
+                                  " different SubDomains currently unsupported")
+    elif len(sdms) == 1:
+        return sdms.pop()
+    return None
+
+
+def adjust_interp_indices(subdomain, mapper, smapper, cdmapper):
```

Comment:
I tend to prefer creating new objects off the input, rather than modifying the input. It's neater and easier to debug should we ever have to